src/main/java/org/spongepowered/api/Tag.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api;
+
+import com.google.common.reflect.TypeToken;
+import org.spongepowered.api.block.BlockTags;
+
+import java.util.Collection;
+
+/**
+ * Represents a tag which can be applied to a specific {@link CatalogType}.
+ *
+ * @param <C> The catalog type
+ * @see BlockTags
+ */
+public interface Tag<C extends CatalogType> extends CatalogType {
+
+    /**
+     * Gets the {@link TypeToken} of the target catalog type.
+     *
+     * @return The target catalog type
+     */
+    TypeToken<C> getTargetType();
```

Comment:
No way to guarantee this either, Tags can be completely ignorant (or empty for that matter) to know what their generic type information would contain. Unless you plan on providing `TypeToken.of(Object.class)` as a worst case scenario.